### PR TITLE
feat(ui): Add metrics powered performance table [INGEST-544]

### PR DESCRIPTION
**Pull request by matejminar**
Adding the metrics-powered performance landing page table.

image

@k-fish @Zylphrex I chose to make a frontend indirection component for this exploration so that the footprint in performance code is minimal while allowing us to test things fully. As we discussed, this logic will be handled in the future by API.

Sorting and some fields are not yet supported. Here's the list of fields that the team is still working on and will therefore display n/a:
```
transaction.op
http.method
failure_rate()
apdex()
count_miserable(user)
user_misery()
team_key_transaction
```

### CODE REVIEW

File: static/app/utils/discover/discoverOrMetricsQuery.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  convertDiscoverFieldsToMetricsGroupBys,
+} from 'sentry/utils/metrics/fields';
+import MetricsRequest from 'sentry/utils/metrics/metricsRequest';
+import {useMetricsSwitch} from 'sentry/views/performance/metricsSwitch';
+
+import withApi from '../withApi';
+
+import DiscoverQuery, {DiscoverQueryPropsWithThresholds} from './discoverQuery';
+
+function DiscoverOrMetricsQuery(props: DiscoverQueryPropsWithThresholds) {
```
Comment (matejminar) [inline static/app/utils/discover/discoverOrMetricsQuery.tsx:12]: This is the indirection component I mentioned in the PR description.

File: static/app/utils/metrics/fields.tsx [hunk 2]
```diff
@@ -59,3 +65,90 @@ export const METRIC_TO_COLUMN_TYPE: Readonly<
...
+  'p95()': `p95(${TransactionMetric.SENTRY_TRANSACTIONS_TRANSACTION_DURATION})`,
+  'p50(transaction.duration)': `p50(${TransactionMetric.SENTRY_TRANSACTIONS_TRANSACTION_DURATION})`,
+  'p75(transaction.duration)': `p75(${TransactionMetric.SENTRY_TRANSACTIONS_TRANSACTION_DURATION})`,
+  'p95(transaction.duration)': `p95(${TransactionMetric.SENTRY_TRANSACTIONS_TRANSACTION_DURATION})`,
+  'p75(measurements.app_start_cold)': `p75(${TransactionMetric.SENTRY_TRANSACTIONS_MEASUREMENTS_APP_START_COLD})`,
+  'p75(measurements.app_start_warm)': `p75(${TransactionMetric.SENTRY_TRANSACTIONS_MEASUREMENTS_APP_START_WARM})`,
+  'p75(measurements.frames_slow_rate)': `p75(${TransactionMetric.SENTRY_TRANSACTIONS_MEASUREMENTS_FRAMES_SLOW_RATE})`,
+  'p75(measurements.frames_frozen_rate)': `p75(${TransactionMetric.SENTRY_TRANSACTIONS_MEASUREMENTS_FRAMES_FROZEN_RATE})`,
+
+  // TODO(metrics):
```
Comment (matejminar) [inline static/app/utils/metrics/fields.tsx:88]: Listed fields will be added to this map as the backend starts supporting them.

File: static/app/utils/metrics/fields.tsx [hunk 3]
```diff
@@ -59,3 +65,90 @@ export const METRIC_TO_COLUMN_TYPE: Readonly<
...
+  transaction: 'transaction',
+
+  // TODO(metrics):
+  // project
+};
+
+const METRIC_TO_DISCOVER_FIELD = invert(DISCOVER_FIELD_TO_METRIC);
+const METRIC_GROUP_BY_TO_DISCOVER_FIELD = invert(DISCOVER_FIELD_TO_METRIC_GROUP_BY);
+
+export function convertDiscoverFieldsToMetrics<T extends string | string[]>(
```
Comment (matejminar) [inline static/app/utils/metrics/fields.tsx:108]: Bunch of helpers to convert metric fields to discover fields and the other way around.
There's a high probability we'll be able to get rid of these once the logic is moved over to the API.

File: static/app/utils/metrics/metricsRequest.tsx [hunk 4]
```diff
@@ -47,11 +53,16 @@ type Props = DefaultProps & {
   limit?: number;
   interval?: string;
   isDisabled?: boolean;
+  /**
+   * Transform the response data to be something ingestible by GridEditable tables
+   */
+  includeTabularData?: boolean;
```
Comment (matejminar) [inline static/app/utils/metrics/metricsRequest.tsx:59]: MetricsRequest will now be able to return not only raw responses but also data pre-massaged for our tables.

File: static/app/views/performance/table.tsx [hunk 5]
```diff
@@ -395,7 +393,7 @@ class _Table extends React.Component<Props, State> {
 
     return (
       <div>
-        <DiscoverQuery
+        <DiscoverOrMetricsQuery
```
Comment (matejminar) [inline static/app/views/performance/table.tsx:396]: This is the only change that has been made to the performance code.

File: static/app/utils/discover/discoverOrMetricsQuery.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    return (
+      <MetricsRequest
+        api={api}
+        orgSlug={orgSlug}
+        start={start}
+        end={end}
+        statsPeriod={statsPeriod}
+        project={project}
+        environment={environment}
+        // query={eventView.getQueryWithAdditionalConditions()} TODO(metrics)
```
Comment (jjbayer) [inline static/app/utils/discover/discoverOrMetricsQuery.tsx:29]: Any reason we do not plug in

```javascript
query={new MutableSearch(eventView.query).formatString()}
```

here, as for other instances of `MetricRequest`?

File: static/app/utils/discover/discoverOrMetricsQuery.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        field={convertDiscoverFieldsToMetrics(fields)}
+        groupBy={convertDiscoverFieldsToMetricsGroupBys(fields)}
+        orderBy={undefined} // TODO(metrics): waiting for api team_key_transactions sorting
```
Comment (jjbayer) [inline static/app/utils/discover/discoverOrMetricsQuery.tsx:32]: Even without `team_key_transactions` working with metrics, can we order by the secondary column selected for sorting here? Then we can see the pagination work done on the backend in action.

File: static/app/utils/discover/discoverOrMetricsQuery.tsx (hunk 6 above)
Comment (matejminar) [reply static/app/utils/discover/discoverOrMetricsQuery.tsx]: Not really, I didn't expect we would need it for testing the POC. I added it there 👍

File: static/app/utils/discover/discoverOrMetricsQuery.tsx (hunk 7 above)
Comment (matejminar) [reply static/app/utils/discover/discoverOrMetricsQuery.tsx:33]: I'll tacke both sorting and pagination in a follow-up PR. Tracking it here in JIRA: 
https://getsentry.atlassian.net/browse/INGEST-871
https://getsentry.atlassian.net/browse/INGEST-872

Comment (jjbayer) [reply static/app/utils/discover/discoverOrMetricsQuery.tsx:33]: OK perfect!

**Review by jjbayer — APPROVED** [review]

**Pull request merged**